Add parsing from xContent to ShardSearchFailure

cbuescher opened the pull request:
In preparation for being able to parse SearchResponse from its rest
representation, this adds fromXContent to ShardSearchFailure.

javanna commented on core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java:
```diff
@@ -0,0 +1,91 @@
+        ShardSearchFailure failure = new ShardSearchFailure(new ParsingException(0, 0, "some message", null),
+                new SearchShardTarget("nodeId", new ShardId(new Index("indexName", "indexUuid"), 123)));
+        BytesReference xContent = toXContent(failure, XContentType.JSON);
+        assertEquals(
```
shall we rather use assertToXContentEquivalent ? I am always cautious on comparing json strings (keys ordering etc.) but maybe in this case it's ok

cbuescher replied:
I think in the toXContent tests that are more a way of checking the intended output is as expected (and more to show a simple, fixed example output in the tests), string comparrison is more intuitive and shouldn't break. We also do this in the other toXContent tests, so I will leave this also to stay consistent.

javanna replied:
no objections :)

javanna reviewed (approved):
left a nit, LGTM

tlrx commented on core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java:
```diff
@@ -0,0 +1,91 @@
+
+        ShardSearchFailure parsed;
+        try (XContentParser parser = createParser(xContentType.xContent(), originalBytes)) {
+            parser.nextToken(); // jump to first START_OBJECT
```
Why not `assertEquals(XContentParser.Token.START_OBJECT, parser.nextToken());` ?

tlrx reviewed (approved):
I left a minor comment in tests, otherwise LGTM

cbuescher committed:
Add parsing from xContent to ShardSearchFailure

In preparation for being able to parse SearchResponse from its rest
representation, this adds fromXContent to ShardSearchFailure.
